### PR TITLE
Add support for Mesa's Gallium "Nine" statetracker

### PR DESCRIPTION
**Pull request by siro20**
**Changes made since PR #303:** 

- Used wine coding style instead of Mesa's coding style

- Cleaned up configure script to check for headers and libs

- Replaced malloc/free where necessary

- Got rid of pthread dependency

- Use single wrapper dll named d3d9-nine.dll

- No more winex11.drv patching required

**The following bugs have been fixed since PR #303:**

- Crash when using DRI2 on 64bits

- D3D9 GetRasterStatus was returning invalid values

- Steamoverlay doesn't work

**The following features have been added since PR #303:**

- Async pixmap presentation for tear-free display with VSYNC disabled

- Support for hotpatching d3d9 functions

- Support for multiple d3d9 devices in one process

- Support for threaded pixmap presentation (required for DRI_PRIME)

**Crosscompiling guidelines:**
To crosscompile d3d9-nine you have to set env PKG_CONFIG_PATH. For example:

> PKG_CONFIG_PATH=/usr/lib/pkgconfig/

**Compiling as module:**
Configure with --with-d3d9-nine-module=manual to prevent d3dadapter9.so detection in configure.
d3d9-nine.dll requires the registry key
`Software\Wine\Direct3DNine\ModulePath`
to be set that specifies the d3dadapter9.so path.

**Using d3d9-nine.dll:**
To use d3d9-nine.dll you have to use the DLLRedirect mechanism, which can be controlled from winecfg. The functionality to enable d3d9-nine isn't part of this PR.

Please review !

### CODE REVIEW

**Comment by darkbasic** [comment]
Please merge it :)

**Comment by whompy** [comment]
What happens in the event that you have both lib32 and lib64 builds in terms of the d3dadapter9.so registry? Is there a means to set the path for both?

**Comment by axeldavy** [comment]
Right, this seems to be a limitation.
Perhaps we should allow setting several paths via Software\Wine\Direct3DNine\ModulePath
That way you'd set the paths for both the 32bits lib and the 64bits lib. If you use 32bits on 64bits, it fails to load and tries another one.

**Comment by qwertychouskie** [comment]
Any word on when this might be merged?

**Comment by qwertychouskie** [comment]
@siro20 Staging has a new patch submission system at https://dev.wine-staging.com/patches/submit/, maybe submitting it there will get it more attention.

**Comment by qwertychouskie** [comment]
@slackner Why haven't we heard anything? @siro20 you may want to merge the top commit from https://github.com/iXit/wine/commits/master into this PR, as it seems to not be included.

**Comment by qwertychouskie** [comment]
@siro20 @slackner Bump?

# ECHO
## echo
### echo
#### echo
echo

**Comment by darkbasic** [comment]
Come on guys, please at least give us a reason why it cannot be merged :(

**Comment by slackner** [comment]
We already had plenty of discussions about the Gallium Nine patchset in the past, and although a lot of things improved since back then, we are still hesitant to add this patchset to Wine Staging. Some of the technical issues are resolved now (*), but there are still various non-technical problems left.

One of my main concerns is that the work is not done by dropping the patchset in Wine Staging. In practice it is more like the opposite, the real work will start afterwards. There will certainly be a lot of bug reports and user questions, but who will take care of them? Most contributors of Wine Staging, including me, work on this project only in their spare time and we don't have much knowledge about the internals of Mesa. I'm also not aware of a any Gallium Nine or Mesa developers actively participating in the Wine project (bugtracker or mailing list).

Although Gallium Nine is split in two parts (Wine and Mesa), most users will only see the Wine part. This is not really surprising when taking into account that Wine is the only project making use of this feature. Our users will therefore expect that they can report all bugs in the WineHQ bug tracker, even for bugs that are caused by the Mesa code. Who is going to take care of those bugs? We would need reliable contributors of Gallium Nine or the Mesa project, who read through all these bugs, decide whether the problem lies in the Mesa or Wine code, etc. So far we haven't seen many active people in the bug tracker, who have enough knowledge about Wine and Mesa to do this job and we can only speculate if it would change after the merge. In the worst case, all we can do is forward users to a different bug tracker. But what is the point of adding the feature in Wine Staging then? Not to mention that we also do no have any control over the Mesa part of the code.

I would like to remind that Wine Staging basically has two goals, a user and a developer oriented one. For developers we provide a way to discuss and test their patches before they are added into the development branch. On the other hand, users get early access to experimental work-in-progress features. In some rare cases we also sometimes add workarounds for bugs that cannot be easily fixed. In this case, both primary goals are not a good motivation to add the Gallium Nine patches. We already know that the current design / architecture will not be accepted by the development branch and to my knowledge Gallium Nine is not necessarily needed in order to fix any particular bugs. The only reason why it is frequently recommended is because it can provide better performance on some hardware.

Adding the patches would only make sense to give ordinary users an easier way to switch to Gallium Nine. In practice this is not going to work though. At least two very popular distributions (Ubuntu and Debian) do not build Mesa with Gallium Nine enabled. Users would still need to add third-party repos, and we certainly can not start building Mesa packages into the WineHQ repository. Another problem is that such distributions usually stick to a stable Mesa release and users won't get any relevant Gallium Nine improvements, even if they download a new Wine build every two weeks. This means that the only way to get a working setup, is to add additional third party repositories. As a result there is basically no difference compared to the current situation. When additional third-party repositories are needed anyway, the corresponding maintainers could also provide patched Wine Stagings builds, for example.

This is just a small collection of thoughts, there are probably more factors, which also play a role. What about DX11 support for example, which also becomes increasingly important? To my knowledge the Gallium Nine people have no plans to support it any time in the near future? If Gallium Nine is really planned as a long term project, why was there never any effort to fix some of the fundamental design issues, which were pointed out in the past? Also, wouldn't it be easier if Gallium Nine was maintained in a single place, either by moving Mesa parts to Wine (Mesa would then only provide a generic interface to access Gallium Nine functions), or the other way around?

TL;DR: From my point of view, there are still several (mainly non-technical) problems left, which make me hesitate to merge this patchset. I am aware that a lot of users are fascinated by the possible speed improvements, and would like to get it merged sooner than later, but I'm not sure how the maintenance and bug reporting should work in practice. Even if we add the patchset, this will not increase chances of it being accepted in the development branch, the feature will still not be available on various popular distributions, and it will have no effect on the progress of Gallium Nine development.

Best regards,
Sebastian

(*) Wine Staging also accepts experimental patches, but from a technical point of view, the implementation is also still far from perfect. Several major design flaws have been made, like not using Wine memory manager functions to allocate interfaces, or directly passing through pointers from non-Wine DLLs which breaks hooking mechanisms.

**Comment by mirh** [comment]
>  In the worst case, all we can do is forward users to a different bug tracker. But what is the point of adding the feature in Wine Staging then? 

Them being able to use the feature in the first place, bugs or not?

> So far we haven't seen many active people in the bug tracker, who have enough knowledge about Wine and Mesa to do this job and we can only speculate if it would change after the merge.

@siro20 @iXit @zhasha @DarkPlayer you up?

>  In practice this is not going to work though. At least two very popular distributions (Ubuntu and Debian) do not build Mesa with Gallium Nine enabled

That most definitively might not happen even assuming this patch ships, but at least for the moment I can 100% see why they don't. 
It's really the usual chicken and egg problem. 

> What about DX11 support for example, which also becomes increasingly important?

Wasn't there some (really basic) d3d10 state tracker in the past?
With no guarantee of even d3d9 being merged though, it's not hard to understand why it went abandoned. 

> We already know that the current design / architecture will not be accepted by the development branch

>  From my point of view, there are still several (mainly non-technical) problems left

These two sentences seems a bit contradicting.

**Comment by darkbasic** [comment]
I confirm that these guys have always been very kind helping tracking down gallium nine bugs.

**Comment by slackner** [comment]
On 21.07.2017 16:35, mirh wrote:
>>  In the worst case, all we can do is forward users to a different bug tracker. But what is the point of adding the feature in Wine Staging then? 
> 
> Them being able to use the feature in the first place, bugs or not?

People already can use this feature. It is just not included in the Wine Staging default set of experimental patches, for which we offer support.

>>  In practice this is not going to work though. At least two very popular distributions (Ubuntu and Debian) do not build Mesa with Gallium Nine enabled
> 
> That most definitively might not happen even assuming this patch ships, but at least for the moment I can 100% see why they don't. 
> It's really the usual chicken and egg problem.

Debian and Ubuntu do not provide Wine Staging packages. As long as there is no change of mind to accept this feature in the development branch, they have no reason to enable it.

> 
>> What about DX11 support for example, which also becomes increasingly important?
> 
> Wasn't there some (really basic) d3d10 state tracker in the past?
> With no guarantee of even d3d9 being merged though, it's not hard to understand why it went abandoned.

To my knowledge, both the d3d9 and d3d10/11 state tracker were written as proof-of-concepts. The original authors have stopped working on it since a long time. The d3d9 state tracker didn't get popular before the Gallium Nine people resurrected this project, unfortunately without fixing the remaining design flaws or discussing a proper design with the Wine project.

> 
>> We already know that the current design / architecture will not be accepted by the development branch
> 
>>  From my point of view, there are still several (mainly non-technical) problems left
> 
> These two sentences seems a bit contradicting. 
> 

The first sentence refers to the development branch, the second to the Staging branch. The people from the development branch already made pretty clear that they will not accept the current solution. For Staging we can decide independently, but the current (unclear) direction makes me question what kind of benefit it will bring.

**Comment by mirh** [comment]
> The people from the development branch already made pretty clear that they will not accept the current solution. 

And ultimately I think that's the only real problem. 
Though, afaik the only *big deal* [seemed](https://www.phoronix.com/scan.php?page=news_item&px=Wine-D3D9-Gallium-Nine-Opposed) to be the amount of "code duplication" and "still quite sub-par performance". 

As for the former, current patch is less then 5 times what was advertised there. 
I don't know how much it could further be squeezed(/moved the burden to mesa) but this already seems a huge difference, worth _at least_ a re-evaluation. 
While for benchmarks they seem [better](https://wiki.ixit.cz/d3d9#faster_than_windows) than anything one could imagine. 

In an ideal world, I think wine would use native state trackers if available, else low overhead wrappers like [VK9](https://github.com/disks86/VK9), otherwise fallback to *universal* GL.

**Comment by leipero** [comment]
> While for benchmarks they seem better than anything one could imagine.

That's even not "best case scenario", with slower CPU's performance difference is even bigger.

Just trowing idea out there, I don't know if it's possible or not, and how hard is to implement, but wouldn't be better to use some sort of "module design" in WINE, so that there wouldn't be need for re-compiling software, where users could just "browse and add" module? That would be useful not only for gallium-nine, but other things also if possible.

**Comment by mirh** [comment]
Please, let's not burry ourself in fantasy developing. 

We'll hear from OP how much further "burden on wine" can be relieved, but there's only so much you can do with low level stuff. 
Also, I'm not sure how much difficult making d3d9 dll just a router would be, but it very likely outshine the effort for any user to just "add PPA, download".

**Comment by slackner** [comment]
On 21.07.2017 19:54, leipero wrote:
>> While for benchmarks they seem better than anything one could imagine.
> 
> That's even not "best case scenario", with slower CPU's performance difference is even bigger.
> 
> Just trowing idea out there, I don't know if it's possible or not, and how hard is to implement, but wouldn't be better to use some sort of "module design" in WINE, so that there wouldn't be need for re-compiling software, where users could just "browse and add" module? That would be useful not only for gallium-nine, but other things also if possible.


Things like that already exist. Wine ships with a GCC wrapper called winegcc, which can be used to compile additional libraries compatible with Wine. In this case, Mesa could use it to provide and ship a d3d9.dll.so compatible with Wine code. All users would have to do is to symlink it into their Wine prefix or add a DLL redirect (or later, maybe just check a checkbox in winecfg). The winelib interface is pretty stable, so (when ignoring the small winex11 patch) the same Mesa build could be used with basically any version of Wine, including stable and development versions. Mesa people should already be aware of this possibility, I suggested it a while back.

**Comment by leipero** [comment]
@slackner So, if I did understand it right, in that case wine could come with just available "ticking box" when compatible mesa with d3d9.dll.so is availalbe on the system independent of wine version?

If so, that is even better solution.

**Comment by slackner** [comment]
On 22.07.2017 00:49, leipero wrote:
> @slackner So, if I did understand it right, in that case wine could come with just available "ticking box" when compatible mesa with d3d9.dll.so is availalbe on the system independent of wine version?
> 
> If so, that is even better solution.


Yes, exactly. With this solution:

- It would be possible to get it working with any version of Wine, not just Wine Staging / custom patched versions. Users just have to enable it when they use a supported driver. Not sure if such a checkbox would be accepted in the development branch, but at least for Staging this is not an issue.

- It would give the Gallium Nine devs full control about all parts of the code. There is no longer any need to be careful about the interface between Mesa and Wine. In fact, merging the Wine and Mesa part might even improve performance because some functions no longer have to be forwarded.

- It would also give Gallium Nine devs the possibility to fix some of the design problems. For programs compiled with winegcc hooking methods should work fine out of the box, and it is possible to get rid of some of the hacks which were added in the past. Also, it would be no problem to switch to Windows allocation functions (HeapAlloc/HeapFree) which is preferred for all Wine stuff. Mixed use of different allocation functions can lead to memory fragmentation and sooner or later to OOM errors.

**Comment by DarkPlayer** [comment]
Regarding performance:

Wine internally uses wined3d to implement everything Direct3D related. The code originaly started with DirectDraw and was extended with each new DirectX version. It now supports Direct3D 1 to Direct3D 11 including the cpu renderer / blitter for DirectDraw. The reason that you see a much better performance with gallium nine is most probably that it contains far less compatibility code. Wine tries to handle very old OpenGL versions and also emulates old Direct3D stuff that is no longer available on modern hardware. There is for example code to handle gpus without NPOT support, a feature which got mandatory in OpenGL 2.0 - 13 years ago. 

All these compatibility checks add a lot of overhead that can affect the performance. I guess that cleaning up wined3d could already improve the efficiency. OpenGL also introduced various extensions in the mean time to reduce the overhead (e.g. direct state access). It is clear that a direct d3d9 state tracker can squeeze out a bit more performance, but I think that wined3d is also still pretty far away from what might be possible.

**Comment by axeldavy** [comment]
A lot has been discussed, and a few clarifications are needed:

. We didn't decide the current scheme of how gallium nine is integrated alone.
The decisions involved discussions with people from our team, mesa and wine.
Packaging a separate wine dll not packaged with wine or wine staging was discussed several times, but was decided against for several reasons.
About the minor change to winex11.drv, as the post says with recent changes "No more winex11.drv patching required". Thus we can totally send a d3d9 dll and use it on any wine. In theory. Practice is that you still need a configure step: decide where the gallium library is, switch the d3d9.dll of wine, etc. This was discussed, but at the time, we thought things were cleared to go into wine staging, and thus we favored that solution. Installing wine-staging is easy for most users. grabing a dll, replace in the correct dirs manually, and setting paths in wine regedit is something much harder for a lot of users. There could be an automatic setup, but our team has no competence in that.

. Yes there are cases where gallium nine works where wine doesn't work at all. For example The Sims2 works on gallium nine, and not on wine.

. About Ubuntu not having gallium nine. That's chicken and egg problem. No package depends on it, why build it ?

. "The d3d9 state tracker didn't get popular before the Gallium Nine people resurrected this project, unfortunately without fixing the remaining design flaws or discussing a proper design with the Wine project." "If Gallium Nine is really planned as a long term project, why was there never any effort to fix some of the fundamental design issues, which were pointed out in the past? " I disagree. We discussed with people from the wine project, and mostly the answer was that we should drop it completly and work on opengl. The initial projects had design flaws, and they were fixed before we merged into Mesa. The design enabled us overtime to add all the features we wanted that required interaction with wine.

. the d3d10/11 project was initiated by different people and had a very different design. This project was dropped from mesa because it wasn't maintained, was only working barely, and was very hard to use (and it depended on winegcc). We don't plan to revive such a project because too much work is needed for non-full time people for this project. If I remember well, the project was the result of a full-time intern for several works at VmWare. A lot more work is needed to make it in a usable state.

. "wouldn't it be easier if Gallium Nine was maintained in a single place, either by moving Mesa parts to Wine (Mesa would then only provide a generic interface to access Gallium Nine functions), or the other way around?"
No, mesa has a lot of utils functions that wine hasn't (and they cannot be copied to wine, license says) that made things way easier. Moreover it isn't possible to make a wrapper around gallium functions, as they change often (and thus mesa devs are against a wrapper). For these reasons, moving parts in Mesa to Wine would be no benefit. The current design has everything that can be in Mesa be in Mesa. Only what has to be in wine is in the dll. (and for the underlying suggestion of building the dll with mesa, the answer is no, because mesa really doesn't want to depend on wine).

**Comment by axeldavy** [comment]
>  but how do users of Debian/Ubuntu obtain the custom Mesa builds, for example?

As I said it's chicken and egg problem. But at least users can do with ppas.

> Why do they reject the dependency on Wine, when the same solution was used for the d3d10/11 state tracker?

This was one of the reason the d3d10/11 state tracker was rejected.

> Anyway, when you no longer require any changes to other parts of Wine, maybe it would really the best idea to provide a separate package then

If anyone has good skills with installers, please help there. We believe however a nice integration in wine staging with a checkbox is easier (and safer, it's hard to trust installers sometimes) for users.

**Comment by mirh** [comment]
> It might be easy to install Wine Staging, but how do users of Debian/Ubuntu obtain the custom Mesa builds, for example?

Didn't you just say that they aren't shipping wine staging packages in the first place?
Considering you'd already have to add a custom PPA to begin with, doing this twice doesn't seem really all that complexity. 

https://launchpad.net/~oibaf/+archive/ubuntu/gallium-nine
https://launchpad.net/~paulo-miguel-dias/+archive/ubuntu/mesa

**Comment by slackner** [comment]
On 22.07.2017 03:17, axeldavy wrote:
> As I said it's chicken and egg problem. But at least users can do with ppas.

I would agree if Wine Staging would be in official Debian/Ubuntu repositories, but it isn't. We have no influence on packaging decisions in the Debian or Ubuntu community. If Gallium Nine can only be used in combination with Wine Staging (as an artifical limition), the chance that these distros will ever provide packages is very small.

**Comment by slackner** [comment]
On 22.07.2017 10:46, mirh wrote:
>> It might be easy to install Wine Staging, but how do users of Debian/Ubuntu obtain the custom Mesa builds, for example?
> 
> Didn't you just say that they aren't shipping wine staging packages in the first place?

I'm talking about official Ubuntu repositories, not about third-party user repositories.

> Considering you'd already have to add a custom PPA to begin with, doing this twice doesn't seem really all that complexity. 
> 
> https://launchpad.net/~oibaf/+archive/ubuntu/gallium-nine
> https://launchpad.net/~paulo-miguel-dias/+archive/ubuntu/mesa

Our repository (which is hosted at dl.winehq.org by the way, so not really an PPA hosted on Launchpad) is exclusively used for packages related to projects associated with WineHQ. Custom Mesa builds wouldn't really fit in there, unfortunately.

**Comment by leipero** [comment]
@slackner Some (not as popular) distributions actually include even staging-nine patched version in repositories, I don't see the reason why Debian/Ubuntu wouldn't do same thing at least for staging. Further, accepting nine in staging would help package maintainers.

Personally, I like the idea of "modular" approach with "checkbox" already integrated in staging, as long as it doesn't put much burden on mesa team.

https://www.archlinux.org/packages/?sort=&arch=x86_64&repo=Multilib&q=wine&maintainer=&flagged=

**Comment by sarnex** [comment]
Hey guys, where are we at with this PR?

Thanks

**Comment by siro20** [comment]
To summarize:
* We want to include a single DLL named d3d9-nine.dll that interfaces mesa's gallium nine. The PR is intended for wine-staging as it provides the DLL redirect feature through winecfg. It does include configure checks and won't build the new DLL until all required libraries and headers are present.

* The whole discussion about packages and distributions doesn't matter. There are lots of parts of wine that do not compile if libraries or headers are missing. Distributions that do not ship mesa-nine will still be able to build wine-staging without doing any modifications.

* Fedora for example already ships mesa-nine and wine-staging.

* I like the idea of a wine subpackage , but that would still require d3d9-nine.dll to be part of wine-staging, or needs to be provided as patch to all packagers. The last one seems more difficult. Packagers would put the d3d9-nine.dll.so into the additional package that would depend on mesa-nine. Not sure how to integrate winecfg menu entry into this extra package...

**Comment by DarkPlayer** [comment]
> I like the idea of a wine subpackage , but that would still require d3d9-nine.dll to be part of wine-staging, or needs to be provided as patch to all packagers. The last one seems more difficult. Packagers would put the d3d9-nine.dll.so into the additional package that would depend on mesa-nine. Not sure how to integrate winecfg menu entry into this extra package...

This is incorrect. The same way as wine can load arbitrary dlls you put into the system32/syswow64 directory, you can also symlink any dll.so file created by winebuild. We also did not refuse to add a winecfg menu entry to make things easier.

**Comment by siro20** [comment]
Still I haven't read any argument why it should be rejected. Of course there are multiple ways of distributing, packaging, symlinking, installing...

According to wine-staging rules, experimental code that adds bugfixes or features, that needs more testing and discussing should be submitted.

Please discuss why to reject this feature.

**Comment by slackner** [comment]
On 18.08.2017 06:52, Patrick Rudolph wrote:
> Still I haven't read any argument why it should be rejected. Of course there are multiple ways of distributing, packaging, symlinking, installing...
> 
> According to wine-staging rules, experimental code that adds bugfixes or features, that needs more testing and discussing should be submitted.
> 
> Please discuss why to reject this feature.
> 


If you haven't seen yet, I already gave a very long explaination in comment https://github.com/wine-compholio/wine-staging/pull/333#issuecomment-317009513. Gallium Nine can not really be compared to other Staging patchsets, and upstream developers already have made clear that they are not fine with the current solution. If we add it anyway, it means that we have the responsibility to keep it working and take care of user requests. Not for some period of time, but "forever", because there are no plans to fix the design.

I can understand that you guys need a better way to distribute Gallium Nine, but from my point of view the packaging aspects are also very important. Whats the point of making it part of Wine Staging when you do not necessarily depend on any Staging feature? How does it help when users report issues with Gallium Nine to us (bug tracker or mailing lists), but we can't really help them because we are not familiar with Mesa internals? Is it really easier for users to setup Gallium Nine on Debian/Ubuntu, when we can not enable it for our WineHQ packages, due to the lack of Nine support in those distros (they also don't support Staging, and have no reason to enable it)?

As Michael already said, adding a checkbox to simplify the usage of Gallium Nine for Staging users is not a problem. When you want to convince us to merge the full patchset, it would really be useful to hear some good reasons why this makes sense. Even when Mesa is not willing to add a dependency on Wine, the second best option would be to write packaging files and provide d3d9-nine as a separate package, similar to how it is done for wineasio [1]. Then it can be used with all versions of Wine, including development and stable versions. It can be updated independent of the Wine version, and maybe it even has a chance to be accepted in official Debian/Ubuntu repos. Isn't that a much better solution?

Best regards,
Sebastian

[1] https://github.com/jhernberg/wineasio

**Comment by mirh** [comment]
> and upstream developers already have made clear that they are not fine with the current solution

AFAIK - and as I posted [above](https://github.com/wine-compholio/wine-staging/pull/333#issuecomment-317036324) - last time upstream expressed its opinion has been almost 3 years ago. 
And basically none of the concerns raised back then exist anymore. 

If you know anything else more up-to-date and actual, please enlighten us.

**Comment by slackner** [comment]
On 18.08.2017 19:12, mirh wrote:
>> and upstream developers already have made clear that they are not fine with the current solution
> 
> AFAIK - and as I posted [above](https://github.com/wine-compholio/wine-staging/pull/333#issuecomment-317036324) - last time upstream expressed its opinion has been almost 3 years ago. 
> And basically none of the concerns raised back then exist anymore. 
> 
> If you know anything else more up-to-date and actual, please enlighten us. 
> 

Maybe I'm wrong, but to my knowledge the concerns of the upstream
developers are still exactly the same (not much has changed).

The first step towards integrating Gallium Nine in upstream Wine would
be to show that such a second backend is really necessary, and that
performance issues and remaining bugs can not be fixed by optimizing
the OpenGL backend. This is a very complex tasks and to my knowledge
noone has tried to do this yet. There are a lot of opportunities to
optimize performance both on the Wine and Mesa side, and adding a new
backend would only make sense if there is still a big difference after
solving all these issues. Even then, if there are certain cases where
OpenGL does not provide all required features, adding a Mesa
extension would probably be preferred instead of adding a secondary
backend.

As a second step, if there is a general agreement that a
Gallium backend is a good idea (lot of people to convince ...), it
would basically be necessary to start again from scratch. Upstream
would never accept a solution which is only compatible with d3d9. What
about ddraw, d3d8, d3d10 or d3d11? The best way to do this would be to
implement a Gallium version of "wined3d" (instead of d3d9). This
means, instead of having a full D3D implementation on the Mesa side,
Mesa would only implement certain required functions, and the rest of
the logic would remain in "wined3d-nine", which is shared between all
d3d* versions. I'm not sure how much code can be reused when doing
such a major rewrite, so its unlikely that Gallium Nine are willing to
do this.

Please note that for Staging those reasons are less important, but
it would still need a convincing reason why a separate package isn't
better (see previous discussions).

Best regards,
Sebastian

**Comment by whompy** [comment]
User chiming in: performance is ludicrously better with gallium nine vs.
wine. I've begun using nouveau with gallium nine on an older nvidia laptop
even. Diablo 3 was basically unplayable on this machine with nvidia blob +
standard wine.

**Comment by leipero** [comment]
@whompy I have to agree with you here, also, can someone with strong CPU test the difference on this track with "High" preset (or even very high) and 16xAF 4xAA? Download Trackmania Nations Forever game (it's free stadium version of Trackmania United Forever, works well in wine). My difference between wine CSMT and gallium nine on slow CPU is 2 times faster with gallium nine (start of the track = 42-45 FPS CSMT, 87-90 FPS nine). Here's the map, there are more heavy maps, but this is good balance for testing IMO.
[24h_VI_SuperStadium_day.Challenge.zip](https://github.com/wine-compholio/wine-staging/files/1235776/24h_VI_SuperStadium_day.Challenge.zip)

On topic for arguments used by upstream developers I did read on some forum/article (forgive me I can't remmember where, maybe phoronix, not sure), argument was "It's not beneficial because it covers small portion of users (meaning mostly AMD users)", I think they should not look at it from that perspective at all if thata rgument still stands, reason = It's not Mesa developers fault that nvidia don't want to get involved in OSS for their GPU's, so rejecting on that basis only helps anti-consumer practise in my opinion. Benefit is clearly there, and it could be benefit for nvidia users also.

**Comment by mirh** [comment]
> Maybe I'm wrong, but to my knowledge the concerns of the upstream developers are still exactly the same (not much has changed).

The "concerns" they still wouldn't permit are `code duplication` and `performance`. 
The thing is, that contrarily to 30 months ago, current state of affairs for gallium-nine isn't all but falling under that umbrella (or if really any, to a **way** diverse extent)
This should at least be acknowledged. Otherwise it's just a broken clock. 

> The first step towards integrating Gallium Nine in upstream Wine would be to show that such a second backend is really necessary, and that performance issues and remaining bugs can not be fixed by optimizing the OpenGL backend.

I, on the other hand, call this a neater approach than translation. Would I be wrong?
(I \*guess\* proof being that two people could get ahead of tens, and in way less time?)

Then, tbh, in my perfect world, you'd have a `native > low-overhead translation > whatever you want to call GL` "backend" querying.. But that's a discussion for another time 

> adding a Mesa extension would probably be preferred instead of adding a secondary backend. 

I'm not sure what this should entail. 
OP already stated (somewhere, don't remember) that he already moved as much code as it could made sense into mesa. 

> What about ddraw, d3d8, d3d10 or d3d11? 

You forget d3d12! /s
...
Seriously is that a point? Will GL translation only be updated whenever it will deemed to be internally perfect? Did anybody figure out that this is the biggest single step they can do without feeling like they are wasting their time hoping for something that will never materialize?
And I'll avoid other snark remarks. 

> The best way to do this would be to implement a Gallium version of "wined3d" (instead of d3d9).

That's pretty... selfish, to say the least. 
Wine shouldn't have any code that doesn't belong to "re-implementing Windows" - and mesa certainly shouldn't have any code aside of pure API handling. 

> but it would still need a convincing reason why a separate package isn't better (see previous discussions)

I have pretty low knowledge of internals, but at least conceptually I think my "ASIO difference" analysis pretty much makes for a pretty good case against that. 
D3D9 *passthrough* is something you should be looking forward with love. 

Nonetheless, I think these last details are quite petty eventually (I mean, I believe OP would be ok with just almost anything as long as this was finally merged). 
In particular when we still have to agree on the "why should we at all" issue. 

---
Also, I'd like to stress people that we don't need performance anecdotes that don't add any particular technical value over what everybody already knew.

**Comment by John-Gee** [comment]
> I'm not sure what this should entail.
OP already stated (somewhere, don't remember) that he already moved as much code as it could made sense into mesa.

I believe that was about new OpenGL extensions needed to get better performance with the standard wined3d approach, it's not related to nine code.

**Comment by mirh** [comment]
> I believe that was about new OpenGL extensions needed to get better performance with the standard wined3d approach, it's not related to nine code.

https://github.com/wine-compholio/wine-staging/pull/333#issuecomment-317138606

**Comment by siro20** [comment]
I'm looking to ship nine in a separate package, but it still needs some tricks to be usable.
The easiest solution I could think of was a new registry setting that would be written in post-install of wine-3rdparty-plugins (like nine). Winecfg would be extended to show a "3rdparty apps" tab where you could enable/disable all the plugins.

The 3rd party package could install a key looking like that:

HKCU/Software/Wine/3rdParty/nine/
> String "Name" = "D3D9 Nine"
> String "Description" = "Redirect D3D9 to Mesa's Gallium Nine. Bypass wined3d's OpenGL backend."
> String "DLL" = "d3d9-nine"
> DWORD DllRedirect = 1
> String DLLRedirectTarget = "d3d9"

**Comment by mirh** [comment]
> This should at least be acknowledged. Otherwise it's just a broken clock.

Friendly remainder that (for as much as OP is banging his head yet again to sidestep even this blockage) the ball is still in your court to re-evaluate Nine for its \***current**\* merits or demerits - and not those of past tries.

**Comment by siro20** [comment]
I was able to create a standalone package that depends on wine-staging.
The SPEC file is here: https://github.com/siro20/wine-spec
I'm looking for package maintainers to build and maintain it.
It would be great if you could get us in contact with them.

Closing the PR as nine will reside as standalone wine add-on.
